extensions/amp-story/0.1/amp-story.js

```diff
@@ -19,34 +19,734 @@
+    this.isBookendActive_ = false;
+
+    this.vsync_.mutate(() => {
+      this.element.classList.remove('i-amphtml-story-bookend-active');
+    });
+  }
+
+
+  /**
+   * Performs a system navigation if it is determined that the specified event
+   * was a click intended for navigation.
+   * @param {!Event} event 'click' event
+   * @private
+   */
+  maybePerformSystemNavigation_(event) {
+    if (!this.isNavigationalClick_(event)) {
+      // If the system doesn't need to handle this click, then we can simply
+      // return and let the event propagate as it would have otherwise.
+      return;
+    }
+
+    event.stopPropagation();
+
+    // TODO(newmuis): This will need to be flipped for RTL.
+    const nextScreenAreaMin = this.element./*OK*/offsetLeft +
+        ((1 - NEXT_SCREEN_AREA_RATIO) * this.element./*OK*/offsetWidth);
+    const nextScreenAreaMax = this.element./*OK*/offsetLeft +
+        this.element./*OK*/offsetWidth;
+
+    if (event.pageX >= nextScreenAreaMin && event.pageX < nextScreenAreaMax) {
+      this.next_();
+    } else if (event.pageX >= this.element./*OK*/offsetLeft &&
+        event.pageX < nextScreenAreaMin) {
+      this.previous_();
+    }
+  }
+
+
+  /**
+   * @return {!Array<!Array<string>>} A 2D array representing lists of pages by
+   *     distance.  The outer array index represents the distance from the
+   *     active page; the inner array is a list of page IDs at the specified
+   *     distance.
+   */
+  getPagesByDistance_() {
+    const distanceMap = this.getPageDistanceMapHelper_(
+        /* distance */ 0, /* map */ {}, this.activePage_.element.id);
+
+    // Transpose the map into a 2D array.
+    const pagesByDistance = [];
+    Object.keys(distanceMap).forEach(pageId => {
+      const distance = distanceMap[pageId];
+      if (!pagesByDistance[distance]) {
+        pagesByDistance[distance] = [];
+      }
+      pagesByDistance[distance].push(pageId);
+    });
+
+    return pagesByDistance;
+  }
+
+  /**
+   * Creates a map of a page and all of the pages reachable from that page, by
+   * distance.
+   *
+   * @param {number} distance The distance that the page with the specified
+   *     pageId is from the active page.
+   * @param {!Object<string, number>} map A mapping from pageId to its distance
+   *     from the active page.
+   * @param {string} pageId The page to be added to the map.
+   * @return {!Object<string, number>} A mapping from page ID to the priority of
+   *     that page.
+   */
+  getPageDistanceMapHelper_(priority, map, pageId) {
+    if (map[pageId] !== undefined && map[pageId] <= priority) {
+      return map;
+    }
+
+    map[pageId] = priority;
+    const page = this.getPageById_(pageId);
+    page.getAdjacentPageIds().forEach(adjacentPageId => {
+      if (map[adjacentPageId] !== undefined
+          && map[adjacentPageId] <= priority) {
+        return;
+      }
+
+      map = this.getPageDistanceMapHelper_(priority + 1, map, adjacentPageId);
```

Comment:
Assignment to `map` (and I suppose the return value) doesn't seem necessary.

Reply:
Marked with a TODO